### DISCUSSION TITLE
Modify Tokenizer to Tokenise Post Codes

### DISCUSSION THREAD

**Discussion by Fourthought** [explosion/spaCy#7502] [Help: Coding & Implementations] [answered]
Hi there

Would someone be able to assist with tokenizing UK postcodes please.
I am using the following regex pattern to detect UK Postcodes: https://regex101.com/r/2g6SZk/1
This regex pattern works on the website, however, it does not seem to work with the following entity linker pattern:

`

    pattern = {
	    "label": "POSTCODE",
	    "pattern": [{
	    	    "TEXT": {
			    "REGEX": "^(([A-Z]{1,2}[0-9][A-Z0-9]?|ASCN|STHL|TDCU|BBND|[BFS]IQQ|PCRN|TKCA) ?[0-9][A-Z]{2}|BFPO ?[0-9]{1,4}|(KY[0-9]|MSR|VG|AI)[ -]?[0-9]{4}|[A-Z]{2} ?[0-9]{2}|GE ?CX|GIR ?0A{2}|SAN ?TA1)$"
    		    }
	    }]
        }
`

This appears to be a tokenization problem since when the postcode is separated by whitespace, it is tokenised into two tokens.

`

    [
    , EC1A, 1BB, 
    , W1A, 0AX, 
    , M1, 1AE, 
    , B33, 8TH, 
    , CR2, 6XH, 
    , DN55, 1PT, 
    , IV21, 2LR,
    ]
`

Is there a way, therefore, to modify the tokeniser to tokenise according to the given postcode regex pattern?

**Comment by polm** [comment]
There might be a way to do this by fiddling around with the tokenizer, but I would generally implement this as a match for multiple tokens, like this:

```
import spacy
from spacy.matcher import Matcher

nlp = spacy.load("en_core_web_sm")
matcher = Matcher(nlp.vocab)
patterns = [
        [{
            "TEXT": { "REGEX": "^(([A-Z]{1,2}[0-9][A-Z0-9]?|ASCN|STHL|TDCU|BBND|[BFS]IQQ|PCRN|TKCA) ?[0-9][A-Z]{2}|BFPO ?[0-9]{1,4}|(KY[0-9]|MSR|VG|AI)[ -]?[0-9]{4}|[A-Z]{2} ?[0-9]{2}|GE ?CX|GIR ?0A{2}|SAN ?TA1)$" }
            }],
        [
            # simplified for this example
            {"TEXT": { "REGEX": "[A-Z]{1,2}[0-9][A-Z0-9]"}},
            {"TEXT": { "REGEX": "[0-9][A-Z][A-Z]"}},
            ],
        ]


matcher.add("POSTCODE", patterns)

texts = [
        "They live at EC1A1BB",
        "They live at EC1A 1BB",
        ]

for text in texts:
    print("-----")
    print(f"text: {text}")
    doc = nlp(text)
    matches = matcher(doc)
    for match_id, start, end in matches:
        string_id = nlp.vocab.strings[match_id]  # Get string representation
        span = doc[start:end]  # The matched span
        print(match_id, string_id, start, end, span.text)
```

Another option that might require the least overall work (since you already have a regex you like) is to use your regex directly on the text string to get character offsets and use that to create a [Span](https://spacy.io/api/span) and then pass that to [Doc.set_ents](https://spacy.io/api/doc#set_ents).

**Comment by Fourthought** [comment, marked as answer]
That's been really helpful Paul, thank you.

For anyone else with the same question, below is a pipeline component which answers the original question. I've got about 25 other rules for matching VRNs, Dates, and phone numbers.

The problem is to correctly tokenise entity before tagging and parsing, otherwise some constituent elements can be mis-labelled and skew the output. This pipeline component, therefore, is placed before the `Tagger` so each token can be properly tokenised. It might be preferrable to do this with the `Tokeniser`, but this method allows for addition match functions to be added to verify each of entities and add custom properties. 

(I'm still getting the hang of using the factory, any advice on how to better declare the component would be appreciated!)

`

	import spacy
	from spacy import Language
	from spacy.matcher import Matcher
	from spacy.util import filter_spans

	patterns = []

	if not Language.has_factory("PostTokeniser"):
		@Language.factory("PostTokeniser", assigns=[], default_config={'patterns': []})
		def create_PostTokenizer(nlp, name, patterns):
			return PostTokeniser(nlp.vocab, patterns=patterns)

	class PostTokeniser:

		"""
		pipeline component for tokenising and verifying:
		- dates
		- phone numbers
		- VRNs
		- Postcodes
		step 1: regex matching of full and partial values
		step 2: match multiple partials
		step 3: verify the token as a date, phone number etc
		step 4: merge partial tokens
		step 5: apply appropriate labels

		REGEX References:
		https://gist.github.com/danielrbradley/7567269
		https://www.regextester.com/104299
		"""

		def __init__(self, vocab, patterns=patterns):

			self.nlp = nlp
			self.matcher = Matcher(vocab, validate=True)
			if patterns:
				for pattern in patterns:
					self.matcher.add(pattern["label"], [pattern["pattern"]])

		def __call__(self, doc):

			matches = self.matcher(doc, as_spans=True)
			with doc.retokenize() as retokenizer:
				for span in filter_spans(matches):

					retokenizer.merge(span)
					doc.ents = list(doc.ents) + [span]

			return doc

	patterns = [
		{"label":"POSTCODE","pattern":[{"TEXT":{"REGEX":"^(([A-Z]{1,2}[0-9][A-Z0-9]?|ASCN|STHL|TDCU|BBND|[BFS]IQQ|PCRN|TKCA) ?[0-9][A-Z]{2}|BFPO ?[0-9]{1,4}|(KY[0-9]|MSR|VG|AI)[ -]?[0-9]{4}|[A-Z]{2} ?[0-9]{2}|GE ?CX|GIR ?0A{2}|SAN ?TA1)$"}}]},
		{"label":"POSTCODE", "pattern": [{"LOWER": {"REGEX": "[a-z]{1,2}[0-9][a-z0-9]"}}, {"LOWER": {"REGEX": "[0-9][a-z][a-z]"}}]},
		{"label": "VRN", "pattern": [{"LOWER": {"REGEX": "^(?!.*\/)([a-z]{2}[0-9]{2}$)"}}, {"LOWER": {"REGEX": "^(?!.*\/)([a-z]{3}$)"}}]},
		{"label": "VRN", "pattern": [{"LOWER": {"REGEX": "^[a-z]{2}[0-9]{2}[a-z]{3}$"}}]}
	]

	texts = [
		"They've lived at EC1A1BB",
		"They live at EC1A 1BB",
		"The car has the registration plate MC52 IOP",
		"The car has the registration plate MC52IOP",
		]

	nlp = spacy.load("en_core_web_sm")
	nlp.add_pipe("PostTokeniser", config={"patterns": patterns}, before = "tagger")

	for text in texts:
            doc = nlp(text)
            for ent in doc.ents:
                print(ent.root, '=>', ent.root.ent_type_)


`
